File: ruby/examples/test_eywa_client.rb
#!/usr/bin/env ruby

require_relative '../lib/eywa'

def test_client
  puts "Starting EYWA Ruby client test...\n"
  
  # Initialize the pipe
  open_pipe
  
  begin
    # Test 1: Logging functions
    info("Testing info logging", { test: "info" })
    warn("Testing warning logging", { test: "warn" })
    error("Testing error logging (not a real error)", { test: "error" })
    debug("Testing debug logging", { test: "debug" })
    trace("Testing trace logging", { test: "trace" })
    exception("Testing exception logging", { test: "exception" })
    
    # Test 2: Custom log with all parameters
    log(
      event: "INFO",
      message: "Custom log with all parameters",
      data: { custom: true },
      duration: 1234,
      coordinates: { x: 10, y: 20 },
      time: Time.now
    )
    
    # Test 3: Report
    report("Test report message", { reportData: "test" })
    
    # Test 4: Task management
    update_task(PROCESSING)
    info("Updated task status to PROCESSING")
    
    # Test 5: Get current task
    task_thread = get_task
    begin
      task = task_thread.value
      info("Retrieved task:", task)
    rescue => e
      warn("Could not get task (normal if not in task context)", { error: e.message })
    end
    
    # Test 6: GraphQL query
    info("Testing GraphQL query...")
    gql_thread = graphql('
      {
        searchUser(_limit: 2) {
          euuid
          name
          type
          active
        }
      }
    ')
    
    begin
      result = gql_thread.value
      info("GraphQL query successful", { 
        resultCount: result.dig("data", "searchUser")&.length || 0 
      })
      
      # Show first user if available
      if result["data"] && result["data"]["searchUser"] && result["data"]["searchUser"][0]
        info("First user:", result["data"]["searchUser"][0])
      end
    rescue => e
      error("GraphQL query failed", { error: e.message })
    end
    
    # Test 7: Constants
    info("Testing constants", {
      SUCCESS: SUCCESS,
      ERROR: ERROR,
      PROCESSING: PROCESSING,
      EXCEPTION: EXCEPTION
    })
    
    # Test 8: GraphQL with variables
    info("Testing GraphQL with variables...")
    query = '
      query GetUsers($limit: Int) {
        searchUser(_limit: $limit) {
          name
          type
        }
      }
    '
    variables = { limit: 1 }
    
    gql_thread2 = graphql(query, variables)
    begin
      result2 = gql_thread2.value
      info("GraphQL with variables successful", { result: result2 })
    rescue => e
      error("GraphQL with variables failed", { error: e.message })
    end
    
    # Test 9: Handler registration
    register_handler("test.echo") do |data|
      info("Received echo request", data)
    end
    
    # Give time for async operations
    sleep(2)
    
    # Test complete
    info("All tests completed successfully!")
    close_task(SUCCESS)
    
  rescue => e
    error("Test failed with unexpected error", {
      error: e.message,
      backtrace: e.backtrace[0..5]
    })
    close_task(ERROR)
  end
end

# Run the test
test_client

File: bb/examples/test_eywa_client.clj
#!/usr/bin/env bb

(require '[babashka.deps :as deps])

;; Add the local source to classpath
(deps/add-classpath "src")

(require '[eywa.client :as eywa]
         '[clojure.core.async :as async])

(defn test-client []
  (println "Starting EYWA Babashka client test...\n")

  ;; Initialize the pipe
  (eywa/start) ; or (eywa/open-pipe)

  (try
    ;; Test 1: Logging functions
    (eywa/info "Testing info logging" {:test "info"})
    (eywa/warn "Testing warning logging" {:test "warn"})
    (eywa/error "Testing error logging (not a real error)" {:test "error"})
    (eywa/debug "Testing debug logging" {:test "debug"})
    (eywa/trace "Testing trace logging" {:test "trace"})
    (eywa/exception "Testing exception logging" {:test "exception"})

    ;; Test 2: Custom log with all parameters
    (eywa/log "INFO" "Custom log with all parameters"
              :data {:custom true}
              :duration 1234
              :coordinates {:x 10 :y 20})

    ;; Test 3: Report
    (eywa/report "Test report message"
                 :data {:reportData "test"})

    ;; Test 4: Task management
    (eywa/update-task eywa/PROCESSING)
    (eywa/info "Updated task status to PROCESSING")

    ;; Test 5: Get current task
    (let [task-chan (eywa/get-task)]
      (async/go
        (try
          (let [task (async/<! task-chan)]
            (if (:error task)
              (eywa/warn "Could not get task (normal if not in task context)"
                         {:error (:error task)})
              (eywa/info "Retrieved task:" (:result task))))
          (catch Exception e
            (eywa/warn "Task retrieval failed" {:error (.getMessage e)})))))

    ;; Give async operations time to complete
    (Thread/sleep 500)

    ;; Test 6: GraphQL query
    (eywa/info "Testing GraphQL query...")
    (let [gql-chan (eywa/graphql "
      {
        searchUser(_limit: 2) {
          euuid
          name
          type
          active
        }
      }")]
      (async/go
        (try
          (let [result (async/<! gql-chan)]
            (if (instance? Exception result)
              (eywa/error "GraphQL query failed" {:error (.getMessage result)})
              (do
                (eywa/info "GraphQL query successful"
                           {:resultCount (count (get-in result ["data" "searchUser"]))})
                ;; Show first user if available
                (when-let [first-user (first (get-in result ["data" "searchUser"]))]
                  (eywa/info "First user:" first-user)))))
          (catch Exception e
            (eywa/error "GraphQL execution failed" {:error (.getMessage e)})))))

    ;; Give async operations more time
    (Thread/sleep 1000)

    ;; Test 7: Constants
    (eywa/info "Testing constants" {:SUCCESS eywa/SUCCESS
                                    :ERROR eywa/ERROR
                                    :PROCESSING eywa/PROCESSING
                                    :EXCEPTION eywa/EXCEPTION})

    ;; Test 8: New graphql API (with separate args)
    (eywa/info "Testing new GraphQL API...")
    (let [variables {:limit 1}
          query "query($limit: Int) { searchUser(_limit: $limit) { name } }"
          gql-chan2 (eywa/graphql query variables)]
      (async/go
        (try
          (let [result (async/<! gql-chan2)]
            (if (instance? Exception result)
              (eywa/error "GraphQL with variables failed" {:error (.getMessage result)})
              (eywa/info "GraphQL with variables successful" {:result result})))
          (catch Exception e
            (eywa/error "GraphQL with variables execution failed"
                        {:error (.getMessage e)})))))

    ;; Final wait for async operations
    (Thread/sleep 1000)

    ;; Test complete
    (eywa/info "All tests completed successfully!")
    (eywa/close-task eywa/SUCCESS)

    (catch Exception e
      (eywa/error "Test failed with unexpected error"
                  {:error (.getMessage e)
                   :stack (with-out-str (.printStackTrace e))})
      (eywa/close-task eywa/ERROR))))

;; Run the test
(test-client)
